From 1c67386cc7dd85d06d7aee5c05d3614566b5027d Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sat, 6 Mar 2010 20:18:04 +0000
Subject: [PATCH] CONFIGURATION-409: HierarchicalINIConfiguration.save() now
 works correctly with sections containing a delimiter character in their name.
 Ported fix to configuration2 branch.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/branches/configuration2_experimental@919840 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration2/INIConfiguration.java      |  7 ++++++-
 .../configuration2/TestINIConfiguration.java  | 19 +++++++++++++++++++
 xdocs/changes.xml                             |  4 ++++
 3 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/src/main/java/org/apache/commons/configuration2/INIConfiguration.java b/src/main/java/org/apache/commons/configuration2/INIConfiguration.java
index f0282c6863..607ab0dfcb 100644
--- a/src/main/java/org/apache/commons/configuration2/INIConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/INIConfiguration.java
@@ -257,15 +257,20 @@ public void save(Writer writer) throws ConfigurationException
         PrintWriter out = new PrintWriter(writer);
         for (String section : getSections())
         {
+            Configuration subset;
             if (section != null)
             {
                 out.print("[");
                 out.print(section);
                 out.print("]");
                 out.println();
+                subset = createSubnodeConfiguration(getSectionNode(section));
+            }
+            else
+            {
+                subset = getSection(null);
             }
 
-            Configuration subset = getSection(section);
             Iterator<String> keys = subset.getKeys();
             while (keys.hasNext())
             {
diff --git a/src/test/java/org/apache/commons/configuration2/TestINIConfiguration.java b/src/test/java/org/apache/commons/configuration2/TestINIConfiguration.java
index 966903a417..25421a8065 100644
--- a/src/test/java/org/apache/commons/configuration2/TestINIConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/TestINIConfiguration.java
@@ -679,4 +679,23 @@ public void testLineContinuationAtEnd() throws ConfigurationException
         assertEquals("Wrong value", "one" + LINE_SEPARATOR, config
                 .getString("section5.continueNoLine"));
     }
+
+    /**
+     * Tests whether a configuration can be saved that contains section keys
+     * with delimiter characters. This test is related to CONFIGURATION-409.
+     */
+    public void testSaveKeysWithDelimiters() throws ConfigurationException
+    {
+        INIConfiguration conf = new INIConfiguration();
+        final String section = "Section..with..dots";
+        conf.addProperty(section + ".test1", "test1");
+        conf.addProperty(section + ".test2", "test2");
+        conf.save(TEST_FILE);
+        conf = new INIConfiguration();
+        conf.load(TEST_FILE);
+        assertEquals("Wrong value (1)", "test1", conf.getString(section
+                + ".test1"));
+        assertEquals("Wrong value (2)", "test2", conf.getString(section
+                + ".test2"));
+    }
 }
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 122a5d7df8..733d7ca848 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -79,6 +79,10 @@
     </release>
 
     <release version="1.7" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-409">
+        HierarchicalINIConfiguration now correctly saves sections whose name
+        contains delimiter characters.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-407">
         Fixed a potential IllegalStateException in HierarchicalINIConfiguration
         that can be thrown when the global section is requested concurrently.
